[Quasant] iP

Quasant opened the pull request:
# Github got Screwllum before HSR 💯 

> He might come one day, but not today.

Ever wanted someone you can ask to help you remember the things you need to do? Someone that is ...
- Handsome
- Genius
- Metallic
- Costs the grand total of ~~1600 stellar jades~~ **free!**

All you need to do, is to 
1. Download the jar file from [here](https://github.com/Quasant/ip/releases/tag/v0.1)
2. Open the application
3. Type in the commands! 

Features:

- [x] Basic functionality
- [x] A beautiful GUI
- [ ] Can be used to clear pure-fiction (coming soon in 4.x)
  
---
Here is the `main` function.

```java 
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Quasant committed:
Renamed all instances of Duke to Alvis. Modified initial greeting and goodbye.

Quasant committed:
Modified Alvis to repeat user input and terminate if 'bye' is entered

Quasant committed:
Added ability to store text and display stored text when needed

Quasant committed:
Add Task.java

Task is represented as Strings

There are attributes aside from description that is tagged to a task and a string is not able to encapsulate those attributes

Let's create a class called Task, which allows us to associate such attributes to an instance of that object. This allows us to use the Task class as a parent class for various types of tasks that might appear in the future

Quasant committed:
Change Task class access modifiers

Quasant committed:
Add the feature to toggle whether a task is done

Quasant committed:
Add inheritance for 3 subclasses of the Task class

Quasant committed:
Add support for tracking three types of tasks

Quasant committed:
Modify test scripts

Quasant committed:
Modify the bot to handle errors

Quasant committed:
Add support for deleting tasks

Quasant committed:
Fix whitespaces bug

Quasant committed:
Change ArrayList name and certain string outputs

Quasant committed:
Remove redundant attribute and utilised String.format()

Quasant committed:
Add functionality to save and load task list

Quasant committed:
Add to gitignore the save file

Quasant committed:
Merge branch-Level-7 into master

Quasant committed:
Remove debugging statements

Quasant committed:
Add LocalDateTime parsing for Event class

Quasant committed:
Add LocalDateTime parsing for Deadline class

Quasant committed:
Merge branch-Level-8 into master

thisisaditya17 commented on src/main/java/Alvis.java:
```diff
@@ -0,0 +1,129 @@
+public class Alvis {
+    private static final String FILE_PATH = "./Save.txt";
+    
+    public static void main(String[] args) throws IOException {
```
Hmmm... 🤔 
Isn't main method too cramped right now?

Suggestion: Consider creating a different method for processing the input instead of putting everything inside main. It can cause issues when you'll be trying to further scale it and will make error handling and testing difficult. 😸 
Can create a method like startChatbot().

thisisaditya17 commented on src/main/java/Alvis.java:
```diff
@@ -0,0 +1,129 @@
+        System.out.println("You have " + taskList.size() + " tasks");
+    }
+
+    private static void loadTasks(List<Task> taskList) throws IOException {
```
Idk if you have reached to that Level in the ip Progress where you'll have to implement better OOP. 
But can separate saving and loading tasks into a different class. 🐱 🐈

thisisaditya17 commented on src/main/java/Alvis.java:
```diff
@@ -0,0 +1,129 @@
+                writeToFile(taskList);
+                break;
+            case "deadline":
+                int byIndex = userInput.indexOf("/by");
```
I feel rn, the main method is doing everything. Must be tiring for it. 😩 

Suggestion: Instead of main method handling everything like taking input and processing it, you can break those tasks into different classes like ProcessTasks.class, SaveAndLoad.class, etc. according to your taste. 🕺🏼

Quasant replied:
Hahaha indeed my bot was getting tired. I have split up the workload into various other classes. Hopefully, the code is still easy to understand!

thisisaditya17 reviewed (commented):
Overall, good code. Easy to understand. 
Just consider my sugesstions. 😄

Quasant committed:
Push branch level 8

Quasant committed:
Move everything into a package

Quasant committed:
Merge branch-Level-8 into master again

Shantaa commented on src/main/java/Alvis/Deadline.java:
```diff
@@ -0,0 +1,23 @@
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
+    private LocalDateTime deadline;
```
Would it better to rename this variable, maybe something like "due"? So that others don't misunderstand this as an instance of the Deadline class

Quasant replied:
That is a good point! I had taken your suggestion and renamed it to `by`.

Shantaa reviewed (commented):
LGTM

Quasant committed:
Move task related classes into the tasks package

Quasant committed:
Rename all instances of Alvis to Screwllum

Quasant committed:
Create Parser, Storage, Ui and TaskList class

Quasant committed:
Rename TaskList to TaskManager

Quasant committed:
Implement Parser class

Quasant committed:
Add Parser class to screwllum package

Quasant committed:
Implement TaskManager class

Quasant committed:
Change error text

Quasant committed:
Implement Ui class and modify TaskManager to utilise it

Quasant committed:
Add overloaded constructor for TaskManager

Quasant committed:
Implement Storage class

Quasant committed:
Create exceptions related to the chatbot

Quasant committed:
Change error text

Quasant committed:
Add methods to parse dates to strings and vice versa

Quasant committed:
Modify tasks with dates to utilise methods from the Parser class

Quasant committed:
Modify TaskManager to throw ScrewllumExceptions

Quasant committed:
Delete TaskList.java

Quasant committed:
Modify Screwllum to utilise other classes

Quasant committed:
Merge branch-MoreOOP into master

Quasant committed:
Move Storage, Parser and Ui into utils

Quasant committed:
Merge remote-tracking branch 'origin/add-gradle-support' into Master

Quasant committed:
Add Gradle to automate build tasks

Quasant committed:
Add ParserTest class

Quasant committed:
Introduce possible thrown Exception during loading of save files

Quasant committed:
Refactor long methods into smaller chunks

Quasant committed:
Add ParserTest and StorageTest

Quasant committed:
Extract helper method from Storage load method

Quasant committed:
Modify build.gradle to run shadowJar

Quasant committed:
Add JavaDocs comments

Quasant committed:
Change ordering of import statements to adhere to coding standards

Quasant committed:
Implement find command

Quasant committed:
Merge branch 'branch-A-CodingStandard'

Quasant committed:
Merge branch 'branch-A-JavaDoc'

Quasant committed:
Merge branch 'branch-Level-9'

Quasant committed:
Add resource directory

Quasant committed:
Fix checkstyle mistakes

Reorder import statements and remove whitespaces at the end of the lines

This is to be in line with the coding standard. However, I did ignore the warning which involves leaving empty lines.

There is no requirement within the coding standard to pack everything together, which makes it hard to tell when one method starts and another method end. It is also in line with Oracle's coding standard.

Refer to https://www.oracle.com/java/technologies/javase/codeconventions-whitespace.html

Quasant committed:
Add images to resource directory

Quasant committed:
Add code from JavaFX tutorial as the starting block for the GUI

Quasant committed:
Modify build.gradle to run Launcher as mainclass and include JavaFX dependencies

Quasant committed:
Add overloaded constructor for Screwllum

Quasant committed:
Modify methods to transition from a CLI to a GUI

Quasant committed:
Modify methods to return a string instead of void

Quasant committed:
Modify methods to return a string instead of void

Quasant committed:
Modify instances of Duke to Screwllum

Quasant committed:
Fix ground-breaking bug

Image userImage loads resources from huehue.png

However, the file was actually saved as huehue.jpg

The fix was simple, but finding the root cause was a good learning experience. The error messages kept pointing at the Main class as the culprit, but going through the stack trace, managed to find out the issue was in the MainWindow class

Quasant committed:
Polish GUI details

Quasant committed:
Merge branch 'branch-Level-10'

Quasant committed:
Remove trailing white-spaces and added more JavaDocs

Quasant committed:
Add support in build.gradle for running assertions

Quasant committed:
Add assertion within Parser class to ascertain that input is not null

Quasant committed:
Add assertion for Ui and TaskManager class that the token list argument is not null

Quasant committed:
Trim long methods into shorter ones by extracting out smaller methods

Quasant committed:
Modify welcome message

Quasant committed:
Modify method names to make the code more readable

Quasant committed:
Modify execute method to reduce LoC

Quasant committed:
Merge pull request #2 from Quasant/branch-A-Assertions

Merge Branch a assertions to master

Quasant committed:
Merge master into branch-A-CodeQuality

Quasant committed:
Merge pull request #3 from Quasant/branch-A-CodeQuality

After merging master into Branch a code quality, able to merge code quality into master.

Quasant committed:
Add support for archive command

Quasant committed:
Add support for archive command

Quasant committed:
Modify JavaDocs for saveTasks method

Quasant committed:
Add support for archive command as well as improved bye response

Quasant committed:
Add support for archive command

Quasant committed:
Modify Screwllum to be able to handle the archive command

Quasant committed:
Fix checkstyle issues in unit tests

Quasant committed:
Merge branch 'branch-BCD-Extension'

Quasant committed:
Add stage title

Quasant committed:
Add new Exception for empty task desc

Quasant committed:
Fix bug where there is an empty task description, removed EmptyDescriptionException

Quasant committed:
Remove EmptyDescriptionException as it's role is filled by InvalidCommandException

Quasant committed:
Add Ui.png

Quasant committed:
Remove unused imports

Quasant committed:
Modify save and archive file location to be within a data folder

Quasant committed:
Modify welcome message depending on if save file exist

Quasant committed:
Implement anchoring for GUI elements

Quasant committed:
Add main.css and dialog-box.css

Quasant committed:
Change CSS to fit the theme

Quasant committed:
Merge branch 'branch-JavaFX-5'

Quasant committed:
Wrote User Guide

Quasant committed:
Fix User Guide

Quasant committed:
Fix User Guide yet again

Quasant committed:
Provide acknowledgements

Quasant committed:
Fix User Guide mistake

Quasant committed:
Modify saving related methods to fix overwriting archive bug

Quasant committed:
Remove unused imports

Quasant committed:
Add more Junit tests
